compiler/noirc_frontend/src/parser/parser/types.rs

```diff
@@ -235,6 +245,23 @@ pub(super) fn slice_type(
 }
 
 pub(super) fn type_expression() -> impl NoirParser<UnresolvedTypeExpression> {
+    type_expression_inner().try_map(UnresolvedTypeExpression::from_expr)
+}
+
+pub(super) fn numeric_generic_type_expression(
+) -> impl NoirParser<(UnresolvedType, Option<Expression>)> {
+    type_expression_inner().map_with_span(|expr, span| {
+        let expr = UnresolvedTypeExpression::from_expr_helper(expr);
+        match expr {
+            Ok(type_expression) => {
+                (UnresolvedTypeData::Expression(type_expression).with_span(span), None)
+            }
+            Err(err_expr) => (UnresolvedType::error(span), Some(err_expr)),
+        }
+    })
+}
+
+fn type_expression_inner() -> impl NoirParser<Expression> {
```

Comment:
Why are we splitting out type expressions from numeric generic type expressions? When would a type expression not be from a numeric generic? We should treat all instances of these the same, including in array and string types where this is still used.

Reply:
> A new parser combinator for numeric generics was introduced to differentiate with the existing type_expression combinator. This is due to usage of or_not in the generic_type_args parser. Due to the or_not combinator the expression error was being ignored. 

From the PR description

Reply:
I renamed to `type_expression_validated` to make this more clear that we want to emit a parser error and not halt the parser in the case of a bad type expression.

Reply:
Wouldn't we still want to use the same parser for here and array/string types though?

Reply:
Confused what you mean? We are using the same parser internally?

Reply:
I was wondering why they didn't both use `type_expression_validated` since they used to call the same `type_expression()` function originally but it is not that important

Reply:
I just didn't want this change to leak into the other parsers that are fine with the parsing halting and do not use `or_not`. I can experiment with using `type_expression_validated` with the array/string types.